### DISCUSSION TITLE
Best way to register array parameters that can default to single scalar

### DISCUSSION THREAD

**Discussion by mcaroba** [libAtoms/QUIP#333] [General]
I am trying to simplify the soap_turbo string and make it more soap-like. What would be the best way to call `param_register()` for a variable that is internally an array, but allowing the user to define it as a scalar (so that all the array elements would default to this scalar)?

Example, in soap_turbo the number of radial basis functions is defined for each element, if you have a soap descriptor with 3 species, you'd do something like `n_max={10 8 9}`. But imagine that you don't want or need to optimize the number of basis functions per species and just want all of them to use the same number by, e.g., defining `n_max=9`. Then the code would set `n_max(1:3) = 9` either inside `param_register()` or in `descriptors.f95` using some "flag" of information retrievable from `param_register()`.

What would be the easiest, most "quipionic" way to go about this?

**Comment by bernstei** [comment]
The only way I can think of right now is a string, say comma separated, which you parse yourself (there might already be a split function you could use).  We could implement a modification to the parser that array parameters can handle handle a single value input, and if so, set all array elements equal to it, but I don't think that happens now.  And to the extent that these strings are expected to be the same format as the 2nd extxyz line (which is not necessarily going to remain true once we have the separate reader, unless we refactor it so it can be used for standalone strings) it'll break the consistency between fortran and python, because only in fortran are the array dimensions known before parsing.

**Comment by mcaroba** [comment]
Reading a string and then doing my own parsing was what I had in mind as fallback option, so I'll probably try that one out. I'll keep an eye for possible issues with quippy.

**Comment by gabor1** [comment]
have a look at how e0 is done - that can be set as a scalar, or as a string (which is then parsed into an array)

-- Gábor

Gábor Csányi
Professor of Molecular Modelling
Engineering Laboratory, University of Cambridge
Pembroke College Cambridge

Pembroke College supports CARA. A Lifeline to Academics at Risk.  http://www.cara.ngo/

**Reply by mcaroba** [reply]
This could be the solution, thanks. I will try to finalize the interface after this grant application frenzy is over, if I haven't lost my head by then!
